## Global Power Plant Database

The Global Power Plant Database is a comprehensive, open source database of power plants around the world. It centralizes power plant data to make it easier to navigate, compare and draw insights for one’s own analysis. The database covers approximately 35,000 power plants from 167 countries and includes thermal plants (e.g. coal, gas, oil, nuclear, biomass, waste, geothermal) and renewables (e.g. hydro, wind, solar). Each power plant is geolocated and entries contain information on plant capacity, generation, ownership, and fuel type. It will be continuously updated as data becomes available.

Key attributes of the database
The database includes the following indicators:

`country` (text): 3 character country code corresponding to the ISO 3166-1 alpha-3 specification 
`country_long` (text): longer form of the country designation
`name` (text): name or title of the power plant, generally in Romanized form
`gppd_idnr` (text): 10 or 12 character identifier for the power plant
`capacity_mw` (number): electrical generating capacity in megawatts
`latitude` (number): geolocation in decimal degrees; WGS84 (EPSG:4326)
`longitude` (number): geolocation in decimal degrees; WGS84 (EPSG:4326)
`primary_fuel` (text): energy source used in primary electricity generation or export
`other_fuel1` (text): energy source used in electricity generation or export
`other_fuel2` (text): energy source used in electricity generation or export
`other_fuel3` (text): energy source used in electricity generation or export
 `commissioning_year` (number): year of plant operation, weighted by unit-capacity when data is available
`owner` (text): majority shareholder of the power plant, generally in Romanized form
`source` (text): entity reporting the data; could be an organization, report, or document, generally in Romanized form
`url` (text): web document corresponding to the `source` field
`geolocation_source` (text): attribution for geolocation information
`wepp_id` (text): a reference to a unique plant identifier in the widely-used PLATTS-WEPP database.
`year_of_capacity_data` (number): year the capacity information was reported
`generation_gwh_2013` (number): electricity generation in gigawatt-hours reported for the year 2013
`generation_gwh_2014` (number): electricity generation in gigawatt-hours reported for the year 2014
`generation_gwh_2015` (number): electricity generation in gigawatt-hours reported for the year 2015
`generation_gwh_2016` (number): electricity generation in gigawatt-hours reported for the year 2016
`generation_gwh_2017` (number): electricity generation in gigawatt-hours reported for the year 2017
`generation_gwh_2018` (number): electricity generation in gigawatt-hours reported for the year 2018
`generation_gwh_2019` (number): electricity generation in gigawatt-hours reported for the year 2019
`generation_data_source` (text): attribution for the reported generation information
`estimated_generation_gwh_2013` (number): estimated electricity generation in gigawatt-hours for the year 2013
`estimated_generation_gwh_2014` (number): estimated electricity generation in gigawatt-hours for the year 2014 
`estimated_generation_gwh_2015` (number): estimated electricity generation in gigawatt-hours for the year 2015 
`estimated_generation_gwh_2016` (number): estimated electricity generation in gigawatt-hours for the year 2016 
`estimated_generation_gwh_2017` (number): estimated electricity generation in gigawatt-hours for the year 2017 
'estimated_generation_note_2013` (text): label of the model/method used to estimate generation for the year 2013
`estimated_generation_note_2014` (text): label of the model/method used to estimate generation for the year 2014 
`estimated_generation_note_2015` (text): label of the model/method used to estimate generation for the year 2015
`estimated_generation_note_2016` (text): label of the model/method used to estimate generation for the year 2016
`estimated_generation_note_2017` (text): label of the model/method used to estimate generation for the year 2017 
Fuel Type Aggregation
We define the "Fuel Type" attribute of our database based on common fuel categories. 

In [1]:
#Importing the necessary libraires
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb
import warnings
warnings.filterwarnings("ignore")

In [2]:
#Reading the dataset
df = pd.read_csv("database_IND.csv")
df

country country_long                          name   gppd_idnr  \
0       IND        India              ACME Solar Tower  WRI1020239   
1       IND        India           ADITYA CEMENT WORKS  WRI1019881   
2       IND        India      AES Saurashtra Windfarms  WRI1026669   
3       IND        India                   AGARTALA GT  IND0000001   
4       IND        India                  AKALTARA TPP  IND0000002   
..      ...          ...                           ...         ...   
902     IND        India                  YERMARUS TPP  IND0000513   
903     IND        India  Yelesandra Solar Power Plant  WRI1026222   
904     IND        India  Yelisirur wind power project  WRI1026776   
905     IND        India                   ZAWAR MINES  WRI1019901   
906     IND        India       iEnergy Theni Wind Farm  WRI1026761   

     capacity_mw  latitude  longitude primary_fuel other_fuel1 other_fuel2  \
0            2.5   28.1839    73.2407        Solar         NaN         NaN   
1           98.0   24.7663    74.6090         Coal         NaN         NaN   
2           39.2   21.9038    69.3732         Wind         NaN         NaN   
3          135.0   23.8712    91.3602          Gas         NaN         NaN   
4         1800.0   21.9603    82.4091         Coal         Oil         NaN   
..           ...       ...        ...          ...         ...         ...   
902       1600.0   16.2949    77.3568         Coal         Oil         NaN   
903          3.0   12.8932    78.1654        Solar         NaN         NaN   
904         25.5   15.2758    75.5811         Wind         NaN         NaN   
905         80.0   24.3500    73.7477         Coal         NaN         NaN   
906         16.5    9.9344    77.4768         Wind         NaN         NaN   

     ...  year_of_capacity_data  generation_gwh_2013 generation_gwh_2014  \
0    ...                    NaN                  NaN                 NaN   
1    ...                    NaN                  NaN                 NaN   
2    ...                    NaN                  NaN                 NaN   
3    ...                 2019.0                  NaN          617.789264   
4    ...                 2019.0                  NaN         3035.550000   
..   ...                    ...                  ...                 ...   
902  ...                 2019.0                  NaN                 NaN   
903  ...                    NaN                  NaN                 NaN   
904  ...                    NaN                  NaN                 NaN   
905  ...                    NaN                  NaN                 NaN   
906  ...                    NaN                  NaN                 NaN   

    generation_gwh_2015 generation_gwh_2016 generation_gwh_2017  \
0                   NaN                 NaN                 NaN   
1                   NaN                 NaN                 NaN   
2                   NaN                 NaN                 NaN   
3            843.747000          886.004428          663.774500   
4           5916.370000         6243.000000         5385.579736   
..                  ...                 ...                 ...   
902            0.994875          233.596650          865.400000   
903                 NaN                 NaN                 NaN   
904                 NaN                 NaN                 NaN   
905                 NaN                 NaN                 NaN   
906                 NaN                 NaN                 NaN   

     generation_gwh_2018  generation_gwh_2019         generation_data_source  \
0                    NaN                  NaN                            NaN   
1                    NaN                  NaN                            NaN   
2                    NaN                  NaN                            NaN   
3             626.239128                  NaN  Central Electricity Authority   
4            7279.000000                  NaN  Central Electricity Authority   
..                   ...    

The dataset has 907 rows and 27 columns

## Understanding the Dataset and EDA 

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 907 entries, 0 to 906
Data columns (total 27 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   country                   907 non-null    object 
 1   country_long              907 non-null    object 
 2   name                      907 non-null    object 
 3   gppd_idnr                 907 non-null    object 
 4   capacity_mw               907 non-null    float64
 5   latitude                  861 non-null    float64
 6   longitude                 861 non-null    float64
 7   primary_fuel              907 non-null    object 
 8   other_fuel1               198 non-null    object 
 9   other_fuel2               1 non-null      object 
 10  other_fuel3               0 non-null      float64
 11  commissioning_year        527 non-null    float64
 12  owner                     342 non-null    object 
 13  source                    907 non-null    object 
 14  url       

As we can observe, the dataset has a mixture of object and float64 datatypes.

We can observe from the above table that many of the columns in the dataset are redundant and dont provide much information to building the prediction model.

Hence, let us drop the respective columns.

In [4]:
df = df.drop(["other_fuel1", "other_fuel2", "other_fuel3", "commissioning_year", "owner", "geolocation_source", "wepp_id", "year_of_capacity_data", "generation_gwh_2013", "generation_gwh_2014", "generation_gwh_2015", "generation_gwh_2016", "generation_gwh_2017", "generation_gwh_2018", "generation_gwh_2019", "generation_data_source", "estimated_generation_gwh"], axis = 1)
df

country country_long                          name   gppd_idnr  \
0       IND        India              ACME Solar Tower  WRI1020239   
1       IND        India           ADITYA CEMENT WORKS  WRI1019881   
2       IND        India      AES Saurashtra Windfarms  WRI1026669   
3       IND        India                   AGARTALA GT  IND0000001   
4       IND        India                  AKALTARA TPP  IND0000002   
..      ...          ...                           ...         ...   
902     IND        India                  YERMARUS TPP  IND0000513   
903     IND        India  Yelesandra Solar Power Plant  WRI1026222   
904     IND        India  Yelisirur wind power project  WRI1026776   
905     IND        India                   ZAWAR MINES  WRI1019901   
906     IND        India       iEnergy Theni Wind Farm  WRI1026761   

     capacity_mw  latitude  longitude primary_fuel  \
0            2.5   28.1839    73.2407        Solar   
1           98.0   24.7663    74.6090         Coal   
2           39.2   21.9038    69.3732         Wind   
3          135.0   23.8712    91.3602          Gas   
4         1800.0   21.9603    82.4091         Coal   
..           ...       ...        ...          ...   
902       1600.0   16.2949    77.3568         Coal   
903          3.0   12.8932    78.1654        Solar   
904         25.5   15.2758    75.5811         Wind   
905         80.0   24.3500    73.7477         Coal   
906         16.5    9.9344    77.4768         Wind   

                                   source  \
0    National Renewable Energy Laboratory   
1                    Ultratech Cement ltd   
2                                     CDM   
3           Central Electricity Authority   
4           Central Electricity Authority   
..                                    ...   
902         Central Electricity Authority   
903   Karnataka Power Corporation Limited   
904                                   CDM   
905                    Hindustan Zinc ltd   
906                                   CDM   

                                                   url  
0    http://www.nrel.gov/csp/solarpaces/project_det...  
1                      http://www.ultratechcement.com/  
2    https://cdm.unfccc.int/Projects/DB/DNV-CUK1328...  
3                               http://www.cea.nic.in/  
4                               http://www.cea.nic.in/  
..                                                 ...  
902                             http://www.cea.nic.in/  
903                          http://karnatakapower.com  
904  https://cdm.unfccc.int/Projects/DB/TUEV-RHEIN1...  
905                           http://www.hzlindia.com/  
906  https://cdm.unfccc.int/Projects/DB/RWTUV134503...  

[907 rows x 10 columns]

As we can see, the dataset now reduces to 907 rows and 10 columns

In [5]:
#Checking for null values in the dataset
df.isnull().sum()

country          0
country_long     0
name             0
gppd_idnr        0
capacity_mw      0
latitude        46
longitude       46
primary_fuel     0
source           0
url              0
dtype: int64

We can see that only longitude and latitude has 46 blank values in the dataset. Hence, let us drop the rows which contain the blank values.

In [6]:
df.dropna(inplace = True)
df

country country_long                          name   gppd_idnr  \
0       IND        India              ACME Solar Tower  WRI1020239   
1       IND        India           ADITYA CEMENT WORKS  WRI1019881   
2       IND        India      AES Saurashtra Windfarms  WRI1026669   
3       IND        India                   AGARTALA GT  IND0000001   
4       IND        India                  AKALTARA TPP  IND0000002   
..      ...          ...                           ...         ...   
902     IND        India                  YERMARUS TPP  IND0000513   
903     IND        India  Yelesandra Solar Power Plant  WRI1026222   
904     IND        India  Yelisirur wind power project  WRI1026776   
905     IND        India                   ZAWAR MINES  WRI1019901   
906     IND        India       iEnergy Theni Wind Farm  WRI1026761   

     capacity_mw  latitude  longitude primary_fuel  \
0            2.5   28.1839    73.2407        Solar   
1           98.0   24.7663    74.6090         Coal   
2           39.2   21.9038    69.3732         Wind   
3          135.0   23.8712    91.3602          Gas   
4         1800.0   21.9603    82.4091         Coal   
..           ...       ...        ...          ...   
902       1600.0   16.2949    77.3568         Coal   
903          3.0   12.8932    78.1654        Solar   
904         25.5   15.2758    75.5811         Wind   
905         80.0   24.3500    73.7477         Coal   
906         16.5    9.9344    77.4768         Wind   

                                   source  \
0    National Renewable Energy Laboratory   
1                    Ultratech Cement ltd   
2                                     CDM   
3           Central Electricity Authority   
4           Central Electricity Authority   
..                                    ...   
902         Central Electricity Authority   
903   Karnataka Power Corporation Limited   
904                                   CDM   
905                    Hindustan Zinc ltd   
906                                   CDM   

                                                   url  
0    http://www.nrel.gov/csp/solarpaces/project_det...  
1                      http://www.ultratechcement.com/  
2    https://cdm.unfccc.int/Projects/DB/DNV-CUK1328...  
3                               http://www.cea.nic.in/  
4                               http://www.cea.nic.in/  
..                                                 ...  
902                             http://www.cea.nic.in/  
903                          http://karnatakapower.com  
904  https://cdm.unfccc.int/Projects/DB/TUEV-RHEIN1...  
905                           http://www.hzlindia.com/  
906  https://cdm.unfccc.int/Projects/DB/RWTUV134503...  

[861 rows x 10 columns]

We see that the dataset now reduces to 861 rows and 10 columns.

It is also easy to observe that columns country and country_long have the same values throughout the dataset. This also does not help us much in building the prediction model.

Hence, let us drop them as well.


In [7]:
df = df.drop(["country", "country_long"], axis = 1)
df

name   gppd_idnr  capacity_mw  latitude  \
0                ACME Solar Tower  WRI1020239          2.5   28.1839   
1             ADITYA CEMENT WORKS  WRI1019881         98.0   24.7663   
2        AES Saurashtra Windfarms  WRI1026669         39.2   21.9038   
3                     AGARTALA GT  IND0000001        135.0   23.8712   
4                    AKALTARA TPP  IND0000002       1800.0   21.9603   
..                            ...         ...          ...       ...   
902                  YERMARUS TPP  IND0000513       1600.0   16.2949   
903  Yelesandra Solar Power Plant  WRI1026222          3.0   12.8932   
904  Yelisirur wind power project  WRI1026776         25.5   15.2758   
905                   ZAWAR MINES  WRI1019901         80.0   24.3500   
906       iEnergy Theni Wind Farm  WRI1026761         16.5    9.9344   

     longitude primary_fuel                                source  \
0      73.2407        Solar  National Renewable Energy Laboratory   
1      74.6090         Coal                  Ultratech Cement ltd   
2      69.3732         Wind                                   CDM   
3      91.3602          Gas         Central Electricity Authority   
4      82.4091         Coal         Central Electricity Authority   
..         ...          ...                                   ...   
902    77.3568         Coal         Central Electricity Authority   
903    78.1654        Solar   Karnataka Power Corporation Limited   
904    75.5811         Wind                                   CDM   
905    73.7477         Coal                    Hindustan Zinc ltd   
906    77.4768         Wind                                   CDM   

                                                   url  
0    http://www.nrel.gov/csp/solarpaces/project_det...  
1                      http://www.ultratechcement.com/  
2    https://cdm.unfccc.int/Projects/DB/DNV-CUK1328...  
3                               http://www.cea.nic.in/  
4                               http://www.cea.nic.in/  
..                                                 ...  
902                             http://www.cea.nic.in/  
903                          http://karnatakapower.com  
904  https://cdm.unfccc.int/Projects/DB/TUEV-RHEIN1...  
905                           http://www.hzlindia.com/  
906  https://cdm.unfccc.int/Projects/DB/RWTUV134503...  

[861 rows x 8 columns]

The dataset now reduces to 861 rows and 8 columns.

## Predicting capacity_mw

In [8]:
df1 = df[["name", "gppd_idnr", "latitude", "longitude", "primary_fuel", "source", "url", "capacity_mw"]]

In [9]:
df1

name   gppd_idnr  latitude  longitude  \
0                ACME Solar Tower  WRI1020239   28.1839    73.2407   
1             ADITYA CEMENT WORKS  WRI1019881   24.7663    74.6090   
2        AES Saurashtra Windfarms  WRI1026669   21.9038    69.3732   
3                     AGARTALA GT  IND0000001   23.8712    91.3602   
4                    AKALTARA TPP  IND0000002   21.9603    82.4091   
..                            ...         ...       ...        ...   
902                  YERMARUS TPP  IND0000513   16.2949    77.3568   
903  Yelesandra Solar Power Plant  WRI1026222   12.8932    78.1654   
904  Yelisirur wind power project  WRI1026776   15.2758    75.5811   
905                   ZAWAR MINES  WRI1019901   24.3500    73.7477   
906       iEnergy Theni Wind Farm  WRI1026761    9.9344    77.4768   

    primary_fuel                                source  \
0          Solar  National Renewable Energy Laboratory   
1           Coal                  Ultratech Cement ltd   
2           Wind                                   CDM   
3            Gas         Central Electricity Authority   
4           Coal         Central Electricity Authority   
..           ...                                   ...   
902         Coal         Central Electricity Authority   
903        Solar   Karnataka Power Corporation Limited   
904         Wind                                   CDM   
905         Coal                    Hindustan Zinc ltd   
906         Wind                                   CDM   

                                                   url  capacity_mw  
0    http://www.nrel.gov/csp/solarpaces/project_det...          2.5  
1                      http://www.ultratechcement.com/         98.0  
2    https://cdm.unfccc.int/Projects/DB/DNV-CUK1328...         39.2  
3                               http://www.cea.nic.in/        135.0  
4                               http://www.cea.nic.in/       1800.0  
..                                                 ...          ...  
902                             http://www.cea.nic.in/       1600.0  
903                          http://karnatakapower.com          3.0  
904  https://cdm.unfccc.int/Projects/DB/TUEV-RHEIN1...         25.5  
905                           http://www.hzlindia.com/         80.0  
906  https://cdm.unfccc.int/Projects/DB/RWTUV134503...         16.5  

[861 rows x 8 columns]

As we can see, it is clear to tell that 5 out of the 8 columns in the dataset contain object datatype.

Let us use Label Encoder and convert them into numericals for building a better prediction model.

In [12]:
#Label encoding object datatypes with Label Encoder
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
for i in list(df.columns):
    if df1[i].dtype=='object':
        df1[i]=le.fit_transform(df1[i])

In [13]:
df1

name  gppd_idnr  latitude  longitude  primary_fuel  source  url  \
0       0        629   28.1839    73.2407             6     109  126   
1       1        492   24.7663    74.6090             1     174  169   
2       2        818   21.9038    69.3732             7      21  196   
3       3          0   23.8712    91.3602             2      22   58   
4       4          1   21.9603    82.4091             1      22   58   
..    ...        ...       ...        ...           ...     ...  ...   
902   856        468   16.2949    77.3568             1      22   58   
903   857        787   12.8932    78.1654             6      77   22   
904   858        846   15.2758    75.5811             7      21  246   
905   859        512   24.3500    73.7477             1      59   82   
906   860        831    9.9344    77.4768             7      21  234   

     capacity_mw  
0            2.5  
1           98.0  
2           39.2  
3          135.0  
4         1800.0  
..           ...  
902       1600.0  
903          3.0  
904         25.5  
905         80.0  
906         16.5  

[861 rows x 8 columns]

Since capacity_mw column consists float64 datatype, the prediction model we have to build is a regression type model.

Let us now proceed in building an appropriate regression type model.

## K-Neighbours Regressor

In [14]:
from sklearn.neighbors import KNeighborsRegressor

In [15]:
from sklearn.model_selection import train_test_split

In [16]:
X1 = df.drop("capacity_mw", axis = 1)
Y1 = df["capacity_mw"]

In [17]:
X1_train, X1_test, Y1_train, Y1_test = train_test_split(X1, Y1, test_size=0.2, random_state=3)

In [19]:
knr = KNeighborsRegressor()
knr.fit(X1_train,Y1_train)
predknr=knr.predict(X1_test)

In [20]:
knr.score(X1_train,Y1_train)

0.3510237027306792

## Random Forest Regressor

In [21]:
from sklearn.ensemble import RandomForestRegressor

In [24]:
rfr=RandomForestRegressor()
rfr.fit(X1_train, Y1_train)
predrfr = rfr.predict(X1_test)

In [25]:
rfr.score(X1_train,Y1_train)

0.9117778277437585

In [26]:
from sklearn.metrics import r2_score

In [28]:
r2_score(Y1_test,predrfc)

0.5143979974725084

We see that Random Forest Regressor Model is providing us with better prediction compared to K-Neighbours Regressor. 

Hence, it is safe to say that we can use Random Forest Regressor to predict the capacity_mw column from the information provided in this dataset.

## Predicting primary_fuel

In [29]:
df2 = df[["name", "gppd_idnr", "latitude", "longitude", "source", "url", "capacity_mw", "primary_fuel"]]
df2

name  gppd_idnr  latitude  longitude  source  url  capacity_mw  \
0       0        629   28.1839    73.2407     109  126          2.5   
1       1        492   24.7663    74.6090     174  169         98.0   
2       2        818   21.9038    69.3732      21  196         39.2   
3       3          0   23.8712    91.3602      22   58        135.0   
4       4          1   21.9603    82.4091      22   58       1800.0   
..    ...        ...       ...        ...     ...  ...          ...   
902   856        468   16.2949    77.3568      22   58       1600.0   
903   857        787   12.8932    78.1654      77   22          3.0   
904   858        846   15.2758    75.5811      21  246         25.5   
905   859        512   24.3500    73.7477      59   82         80.0   
906   860        831    9.9344    77.4768      21  234         16.5   

     primary_fuel  
0               6  
1               1  
2               7  
3               2  
4               1  
..            ...  
902             1  
903             6  
904             7  
905             1  
906             7  

[861 rows x 8 columns]

Predicting the primary_fuel column is to predict the type of primary fuel, and hence, we have to build a classification prediction model to predict the type of primary_fuel.

Let us now build a classification model.

## Support Vector Classifier

In [42]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score,confusion_matrix, classification_report

In [38]:
X2 = df.drop("primary_fuel", axis = 1)
Y2 = df["primary_fuel"]

In [39]:
X2_train, X2_test, Y2_train, Y2_test = train_test_split(X1, Y1, test_size=0.3, random_state=4)

In [40]:
svc = SVC()
svc.fit(X1_train,Y1_train)
predsvc=svc.predict(X1_test)

In [41]:
svc.score(X2_train,Y2_train)

0.7259136212624585

## MultinomialNB

In [45]:
from sklearn.naive_bayes import MultinomialNB

In [46]:
mnb=MultinomialNB()
mnb.fit(X2_train,Y2_train)
predmnb=mnb.predict(X2_test)

In [48]:
mnb.score(X2_train,Y2_train)

0.3953488372093023

As we can observe, Support Vector Classifier does a much better predcition that MultinomialNB in predicting the type of primary_fuel.

Hence, SVC can be used to predict the primary_fuel column better.